builtin/logical/aws/path_roles.go

```diff
@@ -58,94 +85,342 @@ func pathRoles() *framework.Path {
+func legacyRoleData(d *framework.FieldData) (string, error) {
+	policy := d.Get("policy").(string)
+	arn := d.Get("arn").(string)
+
+	switch {
+	case policy == "" && arn == "":
+		return "", nil
+	case policy != "" && arn != "":
+		return "", errors.New("only one of policy or arn should be provided")
+	case policy != "":
+		return policy, nil
+	default:
+		return arn, nil
+	}
+}
+
+func (b *backend) pathRolesWrite(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	var resp logical.Response
+
+	roleName := d.Get("name").(string)
+	if roleName == "" {
+		return logical.ErrorResponse("missing role name"), nil
+	}
+
+	b.roleMutex.Lock()
+	defer b.roleMutex.Unlock()
+	roleEntry, err := nonLockedRoleRead(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		legacyEntry, err := req.Storage.Get(ctx, "policy/"+roleName)
+		if err != nil {
+			return nil, err
+		}
+		if legacyEntry == nil {
+			roleEntry = &awsRoleEntry{}
+		} else {
+			roleEntry = upgradeLegacyPolicyEntry(string(legacyEntry.Value))
+			if roleEntry.InvalidData != "" {
+				resp.AddWarning(fmt.Sprintf("Invalid data of %q cleared out of role", roleEntry.InvalidData))
+				roleEntry.InvalidData = ""
+			}
+			err = nonLockedSetAwsRole(ctx, req.Storage, roleName, roleEntry)
+			if err != nil {
+				return nil, err
+			}
+			err = req.Storage.Delete(ctx, "policy/"+roleName)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	legacyRole, err := legacyRoleData(d)
+	if err != nil {
+		return nil, err
+	}
+
+	if credentialTypeRaw, ok := d.GetOk("credential_type"); ok {
+		if legacyRole != "" {
+			return logical.ErrorResponse("cannot supply deprecated role or policy parameters with an explicit credential_type"), nil
+		}
+		credentialType := credentialTypeRaw.(string)
+		allowedCredentialTypes := []string{iamUserCred, assumedRoleCred, federationTokenCred}
+		if !strutil.StrListContains(allowedCredentialTypes, credentialType) {
+			return logical.ErrorResponse(fmt.Sprintf("unrecognized credential_type: %q, not one of %#v", credentialType, allowedCredentialTypes)), nil
+		}
+		roleEntry.CredentialTypes = []string{credentialType}
+	}
+
+	if roleArnsRaw, ok := d.GetOk("role_arns"); ok {
+		if legacyRole != "" {
+			return logical.ErrorResponse("cannot supply deprecated role or policy parameters with role_arns"), nil
+		}
+		roleEntry.RoleArns = roleArnsRaw.([]string)
+	}
+
+	if policyArnsRaw, ok := d.GetOk("policy_arns"); ok {
+		if legacyRole != "" {
+			return logical.ErrorResponse("cannot supply deprecated role or policy parameters with policy_arns"), nil
+		}
+		roleEntry.PolicyArns = policyArnsRaw.([]string)
+	}
+
+	if policyDocumentRaw, ok := d.GetOk("policy_document"); ok {
+		if legacyRole != "" {
+			return logical.ErrorResponse("cannot supply deprecated role or policy parameters with policy_document"), nil
+		}
+		var compacted string
+		if policyDocumentRaw.(string) == "" {
```

Comment:
You can simplify this with something like this:

```
compacted := policyDocumentRaw.(string)
if len(compacted) > 0 {
...
```

Reply:
Thanks! Done.